axum-extra/src/protobuf.rs

```diff
@@ -0,0 +1,330 @@
+///
+/// # Extractor example
+///
+/// ```rust,no_run
+/// use axum::{routing::post, Router};
+/// use axum_extra::protobuf::ProtoBuf;
+///
+/// #[derive(prost::Message)]
+/// struct CreateUser {
+///     #[prost(string, tag="1")]
+///     email: String,
+///     #[prost(string, tag="2")]
+///     password: String,
+/// }
+///
+/// async fn create_user(ProtoBuf(payload): ProtoBuf<CreateUser>) {
+///     // payload is `CreateUser`
+/// }
+///
+/// let app = Router::new().route("/users", post(create_user));
+/// # async {
+/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
+///
+/// # As response
+///
+/// When used as a response, it can encode any type that implements [`prost::Message`] to
+/// a newly allocated buffer.
+///
+/// If no `Content-Type` header is set, the `Content-Type: application/octet-stream` header
+/// will be used automatically.
+///
+/// # Response example
+///
+/// ```
+/// use axum::{
+///     extract::Path,
+///     routing::get,
+///     Router,
+/// };
+/// use axum_extra::protobuf::ProtoBuf;
+///
+/// #[derive(prost::Message)]
+/// struct User {
+///     #[prost(string, tag="1")]
+///     username: String,
+/// }
+///
+/// async fn get_user(Path(user_id) : Path<String>) -> ProtoBuf<User> {
+///     let user = find_user(user_id).await;
+///     ProtoBuf(user)
+/// }
+///
+/// async fn find_user(user_id: String) -> User {
+///     // ...
+///     # unimplemented!()
+/// }
+///
+/// let app = Router::new().route("/users/:id", get(get_user));
+/// # async {
+/// # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+/// # };
+/// ```
+#[derive(Debug, Clone, Copy, Default)]
+#[cfg_attr(docsrs, doc(cfg(feature = "prost")))]
+pub struct ProtoBuf<T>(pub T);
+
+#[async_trait]
+impl<T, B> FromRequest<B> for ProtoBuf<T>
+where
+    T: Message + Default,
+    B: HttpBody + Send,
+    B::Data: Send,
+    B::Error: Into<BoxError>,
+{
+    type Rejection = ProtoBufRejection;
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let mut bytes = Bytes::from_request(req).await?;
+
+        match T::decode(&mut bytes) {
+            Ok(value) => Ok(ProtoBuf(value)),
+            Err(err) => Err(ProtoBufDecodeError::from_err(err).into()),
+        }
+    }
+}
+
+impl<T> Deref for ProtoBuf<T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<T> DerefMut for ProtoBuf<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl<T> From<T> for ProtoBuf<T> {
+    fn from(inner: T) -> Self {
+        Self(inner)
+    }
+}
+
+impl<T> IntoResponse for ProtoBuf<T>
+where
+    T: Message + Default,
+{
+    fn into_response(self) -> Response {
+        let mut buf = BytesMut::with_capacity(128);
+        match &self.0.encode(&mut buf) {
+            Ok(()) => buf.freeze().into_response(),
```

Comment:
This should work too.

```suggestion
            Ok(()) => buf.into_response(),
```